src/pivot-table/hooks/use-selections-model.ts

```diff
@@ -48,68 +86,68 @@ export default function useSelectionsModel(
   }, [selections, updatePageInfo]);
 
   const isLocked = useCallback(
-    (qType: EngineAPI.NxSelectionCellType, qRow: number, qCol: number) => {
-      switch (qType) {
-        case NxSelectionCellType.NX_CELL_LEFT:
-          return !!selected.find(
-            (cell) =>
-              cell.qType === NxSelectionCellType.NX_CELL_TOP ||
-              (cell.qType === NxSelectionCellType.NX_CELL_LEFT && cell.qCol !== qCol),
-          );
-        case NxSelectionCellType.NX_CELL_TOP:
-          return !!selected.find(
-            (cell) =>
-              cell.qType === NxSelectionCellType.NX_CELL_LEFT ||
-              (cell.qType === NxSelectionCellType.NX_CELL_TOP && cell.qRow !== qRow),
-          );
-        default:
-          return false;
+    (cell: Cell) => {
+      if (selectedPivotCells.size === 0 || selectedRowOrColumn.selectionCellType === null) {
+        return false;
       }
+
+      if (cell.selectionCellType !== selectedRowOrColumn.selectionCellType) {
+        return true;
+      }
+
+      if (cell.selectionCellType === NxSelectionCellType.NX_CELL_LEFT) {
+        return selectedRowOrColumn.coord !== cell.x;
+      }
+
+      return selectedRowOrColumn.coord !== cell.y;
     },
-    [selected],
+    [selectedPivotCells, selectedRowOrColumn],
   );
 
   const select = useCallback(
-    (qType: EngineAPI.NxSelectionCellType, qRow: number, qCol: number) => async (evt: React.MouseEvent) => {
-      if ((evt.target as HTMLElement)?.className.includes("sn-pivot-table-column-adjuster")) {
+    (cell: Cell) => async (evt: React.MouseEvent) => {
+      if ((evt.target as HTMLElement | SVGElement)?.getAttribute("class") === "sn-pivot-table-column-adjuster") {
         return;
       }
 
       if (!selections.isActive()) {
         await selections.begin([Q_PATH]);
       }
 
-      if (isLocked(qType, qRow, qCol)) {
+      if (isLocked(cell)) {
         return;
       }
 
-      setSelected((prev) => {
-        const values = [...prev];
-        const idx = values.findIndex((cell) => cell.qType === qType && cell.qRow === qRow && cell.qCol === qCol);
-        if (idx > -1) {
-          values.splice(idx, 1);
-        } else {
-          values.push({ qType, qRow, qCol });
-        }
-
-        selections
-          .select({
-            method: "selectPivotCells",
-            params: [Q_PATH, values],
-          })
-          .catch(() => {});
-
-        return values;
-      });
+      const { nextSelectedPivotCells, nextSelectedRowOrColumn } = getNextState(
+        cell,
+        selectedPivotCells,
+        selectedRowOrColumn,
+      );
+
+      try {
+        await selections.select({
+          method: "selectPivotCells",
+          params: [
+            Q_PATH,
+            Array.from(nextSelectedPivotCells.values()).map((c) => ({
```

Comment:
```suggestion
            Array.from(nextSelectedPivotCells).map((c) => ({
```
no `.values()` needed

Reply:
also a better suggestion, I think you can even do this:
```suggestion
            Array.from(nextSelectedPivotCells, (c) => ({ 
                qType: c.selectionCellType,
                qRow: c.y,
                qCol: c.x,
             })
            )
```

because `Array.from(...)` accepts a 2nd arg of a transformer

Reply:
Ah yes, good find. I initially used a `Map` instead of `Set`. But did not think of updating these